README.md
[![npm][npm]][npm-url]
[![deps][deps]][deps-url]
[![test][test]][test-url]
[![coverage][cover]][cover-url]
[![chat][chat]][chat-url]

<div align="center">
  <!-- replace with accurate logo e.g from https://worldvectorlogo.com/ -->
  <a href="https://github.com/webpack/webpack">
    <img width="200" height="200" vspace="" hspace="25"
      src="https://cdn.rawgit.com/webpack/media/e7485eb2/logo/icon.svg">
  </a>
  <h1>Node-Zopfli Compression Plugin</h1>
  <p>Node-Zopfli plugin for Webpack.<p>
</div>

<h2 align="center">Install</h2>

```bash
npm i -D zopfli-webpack-plugin
```

<h2 align="center">Usage</h2>

``` javascript
var ZopfliPlugin = require("zopfli-webpack-plugin");
module.exports = {
	plugins: [
		new ZopfliPlugin({
			asset: "[path].gz[query]",
			algorithm: "zopfli",
			test: /\.(js|html)$/,
			threshold: 10240,
			minRatio: 0.8
		})
	]
}
```

<h2 align="center">Arguments</h2>

* `asset`: The target asset name. `[file]` is replaced with the original asset. `[path]` is replaced with the path of the original asset and `[query]` with the query. Defaults to `"[path].gz[query]"`.
* `filename`: A `function(asset)` which receives the asset name (after processing `asset` option) and returns the new asset name. Defaults to `false`.
* `algorithm`: Can be a `function(buf, callback)` or a string. For a string the algorithm is taken from `zopfli`.
* `test`: All assets matching this RegExp are processed. Defaults to every asset.
* `threshold`: Only assets bigger than this size are processed. In bytes. Defaults to `0`.
* `minRatio`: Only assets that compress better that this ratio are processed. Defaults to `0.8`.
* `deleteOriginalAssets`: Whether to delete the original assets or not. Defaults to `false`.

<h2 align="center">Option Arguments</h2>

* verbose: Default: false,
* verbose_more: Default: false,
* numiterations: Default: 15,
* blocksplitting: Default: true,
* blocksplittinglast: Default: false,
* blocksplittingmax: Default: 15

<h2 align="center">Maintainers</h2>

<table>
  <tbody>
    <tr>
      <td align="center">
        <img width="150" height="150"
        src="https://avatars3.githubusercontent.com/u/166921?v=3&s=150">
        </br>
        <a href="https://github.com/bebraw">Juho Vepsäläinen</a>
      </td>
      <td align="center">
        <img width="150" height="150"
        src="https://avatars2.githubusercontent.com/u/8420490?v=3&s=150">
        </br>
        <a href="https://github.com/d3viant0ne">Joshua Wiens</a>
      </td>
      <td align="center">
        <img width="150" height="150"
        src="https://avatars3.githubusercontent.com/u/533616?v=3&s=150">
        </br>
        <a href="https://github.com/SpaceK33z">Kees Kluskens</a>
      </td>
      <td align="center">
        <img width="150" height="150"
        src="https://avatars3.githubusercontent.com/u/3408176?v=3&s=150">
        </br>
        <a href="https://github.com/TheLarkInn">Sean Larkin</a>
      </td>
    </tr>
  <tbody>
</table>


[npm]: https://img.shields.io/npm/v/zopfli-webpack-plugin.svg
[npm-url]: https://npmjs.com/package/zopfli-webpack-plugin

[deps]: https://david-dm.org/webpack-contrib/zopfli-webpack-plugin.svg
[deps-url]: https://david-dm.org/webpack-contrib/zopfli-webpack-plugin

[chat]: https://img.shields.io/badge/gitter-webpack%2Fwebpack-brightgreen.svg
[chat-url]: https://gitter.im/webpack/webpack

[test]: http://img.shields.io/travis/webpack-contrib/zopfli-webpack-plugin.svg
[test-url]: https://travis-ci.org/webpack-contrib/zopfli-webpack-plugin

[cover]: https://codecov.io/gh/webpack-contrib/zopfli-webpack-plugin/branch/master/graph/badge.svg
[cover-url]: https://codecov.io/gh/webpack-contrib/zopfli-webpack-plugin



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.0"></a>
# 0.1.0 (2017-06-24)


### Features

* Add ZopfliPlugin ([55cfda0](https://github.com/webpack-contrib/zopfli-webpack-plugin/commit/55cfda0))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

x.x.x / <year>-<month>-<day>
==================

  * Bug fix -
  * Feature -
  * Chore -
  * Docs -


test/zopfli-webpack-plugin.test.js
import { name as PROJECT_NAME } from '../package.json';
import ZopfliPlugin from '../src';

describe(PROJECT_NAME, () => {
  test('should export the loader', (done) => {
    expect(ZopfliPlugin).toBeInstanceOf(Function);
    done();
  });
});


src/index.js
import url from 'url';
import async from 'async';
import zopfli from 'node-zopfli';
import RawSource from 'webpack-sources/lib/RawSource';

class ZopfliPlugin {
  constructor(options = {}) {
    this.asset = options.asset || '[path].gz[query]';
    this.algorithm = options.algorithm || 'gzip';
    this.filename = options.filename || false;
    this.compressionOptions = {};

    if (typeof this.algorithm === 'string') {
      this.compressionOptions = {
        verbose: hasOwnProperty.call(options, 'verbose') ? options.verbose : false,
        verbose_more: hasOwnProperty.call(options, 'verbose_more') ? options.verbose_more : false,
        numiterations: options.numiterations ? options.numiterations : 15,
        blocksplitting: hasOwnProperty.call(options, 'blocksplitting') ? options.blocksplitting : true,
        blocksplittinglast: hasOwnProperty.call(options, 'blocksplittinglast') ? options.blocksplittinglast : false,
        blocksplittingmax: options.blocksplittingmax ? options.blocksplittingmax : 15,
      };
      this.algorithm = (content, options, fn) => {
        zopfli.gzip(content, options, fn);
      };
    } else if (!this.algorithm) {
      throw new Error('Algorithm incorrect or not found');
    }
    this.test = options.test || options.regExp;
    this.threshold = options.threshold || 0;
    this.minRatio = options.minRatio || 0.8;
    this.deleteOriginalAssets = options.deleteOriginalAssets || false;
  }

  apply(compiler) {
    compiler.plugin('this-compilation', (compilation) => {
      compilation.plugin('optimize-assets', (assets, callback) => {
        async.forEach(Object.keys(assets), (file, cb) => {
          if (Array.isArray(this.test)) {
            if (this.test.every(t => !t.test(file))) {
              return cb();
            }
          } else if (this.test && !this.test.test(file)) {
            return cb();
          }
          const asset = assets[file];
          let content = asset.source();

          if (!Buffer.isBuffer(content)) {
            content = new Buffer(content, 'utf-8');
          }

          const originalSize = content.length;

          if (originalSize < this.threshold) {
            return cb();
          }

          this.algorithm(content, this.compressionOptions, (err, result) => {
            if (err) { return cb(err); }

            if (result.length / originalSize > this.minRatio) { return cb(); }

            const parse = url.parse(file);
            const sub = {
              file,
              path: parse.pathname,
              query: parse.query || '',
            };

            let newFile = this.asset.replace(/\[(file|path|query)\]/g, (p0, p1) => sub[p1]);

            if (typeof this.filename === 'function') {
              newFile = this.filename(newFile);
            }
            assets[newFile] = new RawSource(result);

            if (this.deleteOriginalAssets) {
              delete assets[file];
            }
            cb();
          });
        }, callback);
      });
    });
  }
}

export default ZopfliPlugin;
